File: py/exchangerates_sdk.py
# ExchangeRates SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import ExchangeRatesUtility
from core.spec import ExchangeRatesSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import ExchangeRatesBaseFeature
from features import _make_feature


class ExchangeRatesSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = ExchangeRatesUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return ExchangeRatesUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = ExchangeRatesSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    @property
    def convert(self):
        """Idiomatic facade: client.convert.list() / client.convert.load({"id": ...})."""
        from entity.convert_entity import ConvertEntity
        cached = getattr(self, "_convert", None)
        if cached is None:
            cached = ConvertEntity(self, None)
            self._convert = cached
        return cached

    def Convert(self, data=None):
        # Deprecated: use client.convert instead.
        from entity.convert_entity import ConvertEntity
        return ConvertEntity(self, data)


    @property
    def get_api_root(self):
        """Idiomatic facade: client.get_api_root.list() / client.get_api_root.load({"id": ...})."""
        from entity.get_api_root_entity import GetApiRootEntity
        cached = getattr(self, "_get_api_root", None)
        if cached is None:
            cached = GetApiRootEntity(self, None)
            self._get_api_root = cached
        return cached

    def GetApiRoot(self, data=None):
        # Deprecated: use client.get_api_root instead.
        from entity.get_api_root_entity import GetApiRootEntity
        return GetApiRootEntity(self, data)


    @property
    def get_historical_rate_for_currency_and_date(self):
        """Idiomatic facade: client.get_historical_rate_for_currency_and_date.list() / client.get_historical_rate_for_currency_and_date.load({"id": ...})."""
        from entity.get_historical_rate_for_currency_and_date_entity import GetHistoricalRateForCurrencyAndDateEntity
        cached = getattr(self, "_get_historical_rate_for_currency_and_date", None)
        if cached is None:
            cached = GetHistoricalRateForCurrencyAndDateEntity(self, None)
            self._get_historical_rate_for_currency_and_date = cached
        return cached

    def GetHistoricalRateForCurrencyAndDate(self, data=None):
        # Deprecated: use client.get_historical_rate_for_currency_and_date instead.
        from entity.get_historical_rate_for_currency_and_date_entity import GetHistoricalRateForCurrencyAndDateEntity
        return GetHistoricalRateForCurrencyAndDateEntity(self, data)


    @property
    def get_historical_rates_for_date(self):
        """Idiomatic facade: client.get_historical_rates_for_date.list() / client.get_historical_rates_for_date.load({"id": ...})."""
        from entity.get_historical_rates_for_date_entity import GetHistoricalRatesForDateEntity
        cached = getattr(self, "_get_historical_rates_for_date", None)
        if cached is None:
            cached = GetHistoricalRatesForDateEntity(self, None)
            self._get_historical_rates_for_date = cached
        return cached

    def GetHistoricalRatesForDate(self, data=None):
        # Deprecated: use client.get_historical_rates_for_date instead.
        from entity.get_historical_rates_for_date_entity import GetHistoricalRatesForDateEntity
        return GetHistoricalRatesForDateEntity(self, data)


    @property
    def latest(self):
        """Idiomatic facade: client.latest.list() / client.latest.load({"id": ...})."""
        from entity.latest_entity import LatestEntity
        cached = getattr(self, "_latest", None)
        if cached is None:
            cached = LatestEntity(self, None)
            self._latest = cached
        return cached

    def Latest(self, data=None):
        # Deprecated: use client.latest instead.
        from entity.latest_entity import LatestEntity
        return LatestEntity(self, data)


    @property
    def status(self):
        """Idiomatic facade: client.status.list() / client.status.load({"id": ...})."""
        from entity.status_entity import StatusEntity
        cached = getattr(self, "_status", None)
        if cached is None:
            cached = StatusEntity(self, None)
            self._status = cached
        return cached

    def Status(self, data=None):
        # Deprecated: use client.status instead.
        from entity.status_entity import StatusEntity
        return StatusEntity(self, data)


    @property
    def symbol(self):
        """Idiomatic facade: client.symbol.list() / client.symbol.load({"id": ...})."""
        from entity.symbol_entity import SymbolEntity
        cached = getattr(self, "_symbol", None)
        if cached is None:
            cached = SymbolEntity(self, None)
            self._symbol = cached
        return cached

    def Symbol(self, data=None):
        # Deprecated: use client.symbol instead.
        from entity.symbol_entity import SymbolEntity
        return SymbolEntity(self, data)


    @property
    def timeseries(self):
        """Idiomatic facade: client.timeseries.list() / client.timeseries.load({"id": ...})."""
        from entity.timeseries_entity import TimeseriesEntity
        cached = getattr(self, "_timeseries", None)
        if cached is None:
            cached = TimeseriesEntity(self, None)
            self._timeseries = cached
        return cached

    def Timeseries(self, data=None):
        # Deprecated: use client.timeseries instead.
        from entity.timeseries_entity import TimeseriesEntity
        return TimeseriesEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None):
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk

File: lua/exchange-rates_sdk.lua
-- ExchangeRates SDK

local vs = require("utility.struct.struct")
local Utility = require("core.utility_type")
local Spec = require("core.spec")
local helpers = require("core.helpers")

-- Load utility registration (populates Utility._registrar)
require("utility.register")

-- Load features
local BaseFeature = require("feature.base_feature")
local features_factory = require("features")


local ExchangeRatesSDK = {}
ExchangeRatesSDK.__index = ExchangeRatesSDK


local function _make_feature(name)
  local factory = features_factory[name]
  if factory ~= nil then
    return factory()
  end
  return features_factory.base()
end

ExchangeRatesSDK._make_feature = _make_feature


function ExchangeRatesSDK.new(options)
  local self = setmetatable({}, ExchangeRatesSDK)
  self.mode = "live"
  self.features = {}
  self.options = nil

  local utility = Utility.new()
  self._utility = utility

  local config = require("config")()

  self._rootctx = utility.make_context({
    client = self,
    utility = utility,
    config = config,
    options = options or {},
    shared = {},
  }, nil)

  self.options = utility.make_options(self._rootctx)

  if vs.getpath(self.options, "feature.test.active") == true then
    self.mode = "test"
  end

  self._rootctx.options = self.options

  -- Add features from config.
  local feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
  if feature_opts ~= nil then
    local feature_items = vs.items(feature_opts)
    if feature_items ~= nil then
      for _, item in ipairs(feature_items) do
        local fname = item[1]
        local fopts = helpers.to_map(item[2])
        if fopts ~= nil and fopts["active"] == true then
          utility.feature_add(self._rootctx, _make_feature(fname))
        end
      end
    end
  end

  -- Add extension features.
  local extend = vs.getprop(self.options, "extend")
  if type(extend) == "table" then
    for _, f in ipairs(extend) do
      if type(f) == "table" and type(f.get_name) == "function" then
        utility.feature_add(self._rootctx, f)
      end
    end
  end

  -- Initialize features.
  for _, f in ipairs(self.features) do
    utility.feature_init(self._rootctx, f)
  end

  utility.feature_hook(self._rootctx, "PostConstruct")

  -- #BuildFeatures

  return self
end


function ExchangeRatesSDK:options_map()
  local out = vs.clone(self.options)
  if type(out) == "table" then
    return out
  end
  return {}
end


function ExchangeRatesSDK:get_utility()
  return Utility.copy(self._utility)
end


function ExchangeRatesSDK:get_root_ctx()
  return self._rootctx
end


function ExchangeRatesSDK:prepare(fetchargs)
  local utility = self._utility

  fetchargs = fetchargs or {}

  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "prepare",
    ctrl = ctrl,
  }, self._rootctx)

  local options = self.options

  local path = vs.getprop(fetchargs, "path") or ""
  if type(path) ~= "string" then path = "" end

  local method = vs.getprop(fetchargs, "method") or "GET"
  if type(method) ~= "string" then method = "GET" end

  local params = helpers.to_map(vs.getprop(fetchargs, "params")) or {}
  local query = helpers.to_map(vs.getprop(fetchargs, "query")) or {}

  local headers = utility.prepare_headers(ctx)

  local base = vs.getprop(options, "base") or ""
  if type(base) ~= "string" then base = "" end
  local prefix = vs.getprop(options, "prefix") or ""
  if type(prefix) ~= "string" then prefix = "" end
  local suffix = vs.getprop(options, "suffix") or ""
  if type(suffix) ~= "string" then suffix = "" end

  ctx.spec = Spec.new({
    base = base,
    prefix = prefix,
    suffix = suffix,
    path = path,
    method = method,
    params = params,
    query = query,
    headers = headers,
    body = vs.getprop(fetchargs, "body"),
    step = "start",
  })

  -- Merge user-provided headers.
  local uh = vs.getprop(fetchargs, "headers")
  if type(uh) == "table" then
    for k, v in pairs(uh) do
      ctx.spec.headers[k] = v
    end
  end

  local _, err = utility.prepare_auth(ctx)
  if err ~= nil then
    return nil, err
  end

  return utility.make_fetch_def(ctx)
end


function ExchangeRatesSDK:direct(fetchargs)
  local utility = self._utility

  local fetchdef, err = self:prepare(fetchargs)
  if err ~= nil then
    return { ok = false, err = err }, nil
  end

  fetchargs = fetchargs or {}
  local ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl")) or {}

  local ctx = utility.make_context({
    opname = "direct",
    ctrl = ctrl,
  }, self._rootctx)

  local url = fetchdef["url"] or ""
  local fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

  if fetch_err ~= nil then
    return { ok = false, err = fetch_err }, nil
  end

  if fetched == nil then
    return {
      ok = false,
      err = ctx:make_error("direct_no_response", "response: undefined"),
    }, nil
  end

  if type(fetched) == "table" then
    local status = helpers.to_int(vs.getprop(fetched, "status"))
    local headers = vs.getprop(fetched, "headers") or {}

    -- No-body responses (204, 304) and explicit zero content-length
    -- must skip JSON parsing — calling json() on an empty body errors.
    local content_length = nil
    if type(headers) == "table" then
      content_length = headers["content-length"]
    end
    local no_body = status == 204 or status == 304 or tostring(content_length) == "0"

    local json_data = nil
    if not no_body then
      local jf = vs.getprop(fetched, "json")
      if type(jf) == "function" then
        local ok, result = pcall(jf)
        if ok then
          json_data = result
        end
        -- Non-JSON body: json_data stays nil, status/headers preserved.
      end
    end

    return {
      ok = status >= 200 and status < 300,
      status = status,
      headers = headers,
      data = json_data,
    }, nil
  end

  return {
    ok = false,
    err = ctx:make_error("direct_invalid", "invalid response type"),
  }, nil
end



-- Idiomatic facade: client:Convert():list() / client:Convert():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:Convert(data)
  local EntityMod = require("entity.convert_entity")
  if data == nil then
    if self._convert == nil then
      self._convert = EntityMod.new(self, nil)
    end
    return self._convert
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:GetApiRoot():list() / client:GetApiRoot():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:GetApiRoot(data)
  local EntityMod = require("entity.get_api_root_entity")
  if data == nil then
    if self._get_api_root == nil then
      self._get_api_root = EntityMod.new(self, nil)
    end
    return self._get_api_root
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:GetHistoricalRateForCurrencyAndDate():list() / client:GetHistoricalRateForCurrencyAndDate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:GetHistoricalRateForCurrencyAndDate(data)
  local EntityMod = require("entity.get_historical_rate_for_currency_and_date_entity")
  if data == nil then
    if self._get_historical_rate_for_currency_and_date == nil then
      self._get_historical_rate_for_currency_and_date = EntityMod.new(self, nil)
    end
    return self._get_historical_rate_for_currency_and_date
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:GetHistoricalRatesForDate():list() / client:GetHistoricalRatesForDate():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:GetHistoricalRatesForDate(data)
  local EntityMod = require("entity.get_historical_rates_for_date_entity")
  if data == nil then
    if self._get_historical_rates_for_date == nil then
      self._get_historical_rates_for_date = EntityMod.new(self, nil)
    end
    return self._get_historical_rates_for_date
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Latest():list() / client:Latest():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:Latest(data)
  local EntityMod = require("entity.latest_entity")
  if data == nil then
    if self._latest == nil then
      self._latest = EntityMod.new(self, nil)
    end
    return self._latest
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Status():list() / client:Status():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:Status(data)
  local EntityMod = require("entity.status_entity")
  if data == nil then
    if self._status == nil then
      self._status = EntityMod.new(self, nil)
    end
    return self._status
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Symbol():list() / client:Symbol():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:Symbol(data)
  local EntityMod = require("entity.symbol_entity")
  if data == nil then
    if self._symbol == nil then
      self._symbol = EntityMod.new(self, nil)
    end
    return self._symbol
  end
  return EntityMod.new(self, data)
end


-- Idiomatic facade: client:Timeseries():list() / client:Timeseries():load({ id = ... })
-- Entity access is capitalised (PascalCase) for parity with the other SDKs.
function ExchangeRatesSDK:Timeseries(data)
  local EntityMod = require("entity.timeseries_entity")
  if data == nil then
    if self._timeseries == nil then
      self._timeseries = EntityMod.new(self, nil)
    end
    return self._timeseries
  end
  return EntityMod.new(self, data)
end




function ExchangeRatesSDK.test(testopts, sdkopts)
  sdkopts = sdkopts or {}
  sdkopts = vs.clone(sdkopts)
  if type(sdkopts) ~= "table" then
    sdkopts = {}
  end

  testopts = testopts or {}
  testopts = vs.clone(testopts)
  if type(testopts) ~= "table" then
    testopts = {}
  end
  testopts["active"] = true

  vs.setpath(sdkopts, "feature.test", testopts)

  local sdk = ExchangeRatesSDK.new(sdkopts)
  sdk.mode = "test"

  return sdk
end


return ExchangeRatesSDK
